src/todt.c

```diff
@@ -481,11 +481,13 @@ dt_t **Expression_toDt(Expression *e, dt_t **pdt)
                         pdt = dtnzeros(pdt, vd->offset - offset);
                     e = (*se->elements)[k];
 
+                    dt_t *dt = NULL;
                     Type *tb = vd->type->toBasetype();
                     if (tb->ty == Tsarray)
-                        toDtElem(((TypeSArray *)tb), pdt, e);
+                        toDtElem(((TypeSArray *)tb), &dt, e);
                     else
-                        Expression_toDt(e, pdt);           // convert e to an initializer dt
+                        Expression_toDt(e, &dt);           // convert e to an initializer dt
+                    pdt = dtcat(pdt, dt);
```

Comment:
Well, that may improve many operations in `toDtElem` and `Expression_toDt`, but `dtcat` still remains `O(N^2)`.
A better solution would be to use an array or at least a tailq list for dt, [Issue 14816 – improve dt_t data type for faster appending (tail list or array)](https://issues.dlang.org/show_bug.cgi?id=14816).